task-forage.lic

```diff
@@ -0,0 +1,621 @@
+
+    @boosted = nil
+    @number = nil
+    @item = nil
+    @item_location = nil
+    @tasks_completed = 0
+    @tasks_failed = 0
+
+    echo 'UserVars:' if @debug
+    echo UserVars.task_forage.to_yaml if @debug
+    echo 'forage_settings:' if @debug
+    echo @forage_settings.to_yaml if @debug
+
+    while determine_task_necessary
+      find_giver
+      until get_task
+        # loop until we have an appropriate foraging task
+      end
+      gather_items
+      find_giver
+      complete_task
+    end
+  end
+
+  def follow_wanderer
+    echo 'follow_wanderer' if @debug
+    echo "giver-move flag: #{Flags['giver-move']}" if @debug
+    return unless Flags['giver-move']
+
+    direction = nil
+    @move_regex.each do |regex|
+      direction = Regexp.last_match[1] if Flags['giver-move'].to_s =~ regex
+    end
+    Flags.reset('giver-move')
+    return false unless direction
+
+    return false unless XMLData.room_exits.include?(direction)
+
+    move(direction)
+    true
+  end
+
+  def find_giver
+    echo 'find_giver' if @debug
+    if @task_givers[@task_town]['location'].instance_of? Hash # If the NPC is a wanderer...
+      if Flags['giver-move'] # Did we see the task giver wander away?
+        follow_wanderer
+        search_routine unless check_giver
+      elsif !UserVars.task_forage['npcs'][@task_giver].nil? # Check the last seen location (if known)
+        echo 'Traveling to last known location' if @debug
+        DRCT.walk_to(UserVars.task_forage['npcs'][@task_giver])
+        search_routine unless check_giver
+      else
+        echo 'No last known location. Starting from scratch.' if @debug
+        search_routine
+        check_giver
+      end
+    elsif @task_givers[@task_town]['location'].instance_of? String # If the NPC is not a wanderer, head straight there
+      echo 'Stationary giver. Heading straight there.' if @debug
+      DRCT.walk_to(@task_givers[@task_town]['location'])
+      unless check_giver
+        DRC.message("ERROR: Unable to locate stationary NPC giver #{@task_giver}.")
+        safe_exit
+      end
+    end
+  end
+
+  # To find wandering task givers, there are two methods.
+  # The preferred method is a "compiled" route, speeding you through every possible location in the city, and looking for the NPC.
+  # This is similar to find.lic - but that does a "go2" in every single room, which is slow and unnecessary (unless Shard's streets change layout)
+  # If the compiled route doesn't work for some reason, then the back-up is to use find.lic
+  def search_routine
+    echo 'search_routine' if @debug
+    find_result = fast_find(@task_givers[@task_town]['location'])
+    return if find_result
+
+    echo 'old find' if @debug
+    DRC.wait_for_script_to_complete('find', [@task_town, @task_giver])
+    return if check_giver
+
+    DRC.message('Unable to locate the wandering task giver.')
+    safe_exit
+  end
+
+  def in_room?(npc)
+    DRRoom.npcs.any? { |room_npc| room_npc =~ /\b#{npc}\b/i }
```

Comment:
DRRoom.npcs.include?(npc.capitalize) is probably faster, and since you only use it once, probably doesn't need it's own method.

Reply:
I pulled that function from find.lic. But yeah, you're right.